keps/sig-scheduling/4187-a-generic-scheduler-cache-extension/README.md

```diff
@@ -0,0 +1,777 @@
+
+There are emerging requirements to guarantee critical workloads’ disk IO performance. Current K8S scheduler only provides CPU/RAM and ephemeral storage aware scheduling, but it does not provide disk io resource scheduling. When we use existing scheduler framework to develop the disk IO aware scheduler plugin, we need to  create a separate cache in the scheduler plugin to save the disk IO related information such as available capacity and re-implement the pod/node events handling logic. Due to the independent event handling processes, different scheduling plugins (e.g. in-tree scheduling plugin and the customized out-of-tree scheduling plugin) may make the scheduling decision based on resource information snapshotted at different time points, which may result in unnecessary scheduling failure. For instance, right after pod1 is created on node1, pod1 is deleted. Suppose that the in-tree NodeResourcesFit plugin gets the deletion notification at time t1 and the out-of-tree disk io plugin gets the deletion notification at time t2. It could happen that the two scheduler plugins make their scheduling decision at a time point between t1 and t2. In this case, the resource availability information based on which the two scheduler plugins make their decisions would diverge, which may result in scheduling failure or unnecessary pod preemption. With the cache extension, the first-class resource used by NodeResourcesFit plugin and the io resource used by disk io plugin are updated through one event notification and there will be no divergence of resource information. This reduces the probability of pod rescheduling or preemption.
+
+#### Resource Sharing Among Multiple Scheduler Plugins
+
+In existing k8s, the first-class resources are saved in the scheduler cache. Therefore the k8s in-tree scheduler plugins and the out-of-tree scheduler plugins can both access those resources. If we create an extension of the scheduler cache, then those vendor-specific resources saved in the cache extension can also be shared by multiple out-of-tree scheduler plugins.
+
+### Risks and Mitigations
+
+## Design Details
+
+### Data Structure and Interface
+
+The class diagram below describes the new data structures, modification of the existing data structures, and the new interfaces including **their attributes, operations (or methods), and the relationships** among them. 
+![Class Diagram](class-diagram.png)
+
+#### New Interfaces
+
+- The ExtendedResource interface has been defined to manage the resource capacity/attributes, customize resource calculation model. The AddPod, RemovePod methods are invoked when a Pod is created on a node or removed from a node.  The AdmitPod method is invoked when a plugin checks whether the resource fits for a pod during the filtering stage. Scheduler plugin developers should implement this interface (e.g. XXXResource) to customize their extended resource availability calculation and check of resource fit for a new pod .
```

Comment:
Is the out-of-plugin will call and implemented `AdmitPod` method invocation, this looks like a different model compared with current implemented in scheduler. Scheduler plugins take a set of responsibilities to check the rule for the coming pods, it's not required to implement some interfaces which is bound to a specific resource. Cannot this be done in the filter instead?